fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Utils.java

```diff
@@ -536,4 +569,273 @@ private static <T extends ScalarOperator> void collect(ScalarOperator root, Clas
+            return false;
+        }
+        LogicalAggregationOperator agg = (LogicalAggregationOperator) operator;
+        if (agg.getType() != AggType.GLOBAL) {
+            return false;
+        }
+
+        OptExpression child = root.inputAt(0);
+        return isLogicalSPJ(child);
+    }
+
+    public static boolean isLogicalSPJ(OptExpression root) {
+        if (root == null) {
+            return false;
+        }
+        Operator operator = root.getOp();
+        if (!(operator instanceof LogicalOperator)) {
+            return false;
+        }
+        if (!(operator instanceof LogicalScanOperator)
+                && !(operator instanceof LogicalProjectOperator)
+                && !(operator instanceof LogicalFilterOperator)
+                && !(operator instanceof LogicalJoinOperator)) {
+            return false;
+        }
+        for (OptExpression child : root.getInputs()) {
+            if (!isLogicalSPJ(child)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public static Pair<OptExpression, LogicalPlan> getRuleOptimizedLogicalPlan(String sql,
+                                                                               ColumnRefFactory columnRefFactory,
+                                                                               ConnectContext connectContext) {
+        StatementBase mvStmt;
+        try {
+            List<StatementBase> statementBases =
+                    com.starrocks.sql.parser.SqlParser.parse(sql, connectContext.getSessionVariable());
+            Preconditions.checkState(statementBases.size() == 1);
+            mvStmt = statementBases.get(0);
+        } catch (ParsingException parsingException) {
+            LOG.warn("parse sql:{} failed", sql, parsingException);
+            return null;
+        }
+        Preconditions.checkState(mvStmt instanceof QueryStatement);
+        Analyzer.analyze(mvStmt, connectContext);
+        QueryRelation query = ((QueryStatement) mvStmt).getQueryRelation();
+        LogicalPlan logicalPlan =
+                new RelationTransformer(columnRefFactory, connectContext).transformWithSelectLimit(query);
+        // optimize the sql by rule and disable rule based materialized view rewrite
+        OptimizerConfig optimizerConfig = new OptimizerConfig(OptimizerConfig.OptimizerAlgorithm.RULE_BASED);
+        // TODO: disable mv query rewrite rules
+        Optimizer optimizer = new Optimizer(optimizerConfig);
+        OptExpression optimizedPlan = optimizer.optimize(
+                connectContext,
+                logicalPlan.getRoot(),
+                new PhysicalPropertySet(),
+                new ColumnRefSet(logicalPlan.getOutputColumn()),
+                columnRefFactory);
+        return Pair.create(optimizedPlan, logicalPlan);
+    }
+
+    public static void updatePartialPartitionPredicate(MaterializedView mv, ColumnRefFactory columnRefFactory,
```

Comment:
Plz add some comment, and maybe put it in the MaterializedViewPlanner is better.